server/remote_cache/content_addressable_storage_server/content_addressable_storage_server.go

```diff
@@ -789,53 +812,115 @@ func (s *ContentAddressableStorageServer) GetTree(req *repb.GetTreeRequest, stre
 
 		allDescendents := make([]*capb.DirectoryWithDigest, 0, len(children))
 		allDescendents = append(allDescendents, dirWithDigest)
+		allCachedSubtrees := make([]*digest.ResourceName, 0)
+		allCachedSubtreeContents := make([]*capb.DirectoryWithDigest, 0)
 
 		eg, egCtx := errgroup.WithContext(ctx)
 		for _, childDirWithDigest := range children {
 			childDirWithDigest := childDirWithDigest
 			l := level
 			eg.Go(func() error {
-				grandChildren, err := fetch(egCtx, childDirWithDigest, l+1)
+				grandchild, err := fetch(egCtx, childDirWithDigest, l+1)
 				if err != nil {
 					return err
 				}
 				mu.Lock()
 				defer mu.Unlock()
-				allDescendents = append(allDescendents, grandChildren...)
+				if *getTreeSubtreeSupport && grandchild.cachedRootCASResourceName != nil && req.GetSendCachedSubtreeDigests() && len(grandchild.nonSubtreeDirectories) >= *getTreeSubtreeMinDirCount {
+					allCachedSubtrees = append(allCachedSubtrees, grandchild.cachedRootCASResourceName)
+					allCachedSubtreeContents = append(allCachedSubtreeContents, grandchild.nonSubtreeDirectories...)
+				} else {
+					allDescendents = append(allDescendents, grandchild.nonSubtreeDirectories...)
+					if len(grandchild.cachedSubtrees) > 0 {
+						allCachedSubtrees = append(allCachedSubtrees, grandchild.cachedSubtrees...)
+						allCachedSubtreeContents = append(allCachedSubtreeContents, grandchild.subtreeDirectories...)
+					}
+				}
```

Comment:
could you add a comment on these two cases? I mostly am unsure about the else statement -- even tho subtrees are not requested, we're still appending to allCachedSubtrees and allCachedSubtreeContents?

Reply:
hopefully this is enough commentary